llama.cpp

```diff
@@ -9220,6 +9245,10 @@ struct llama_model * llama_load_model_from_file(
     return model;
 }
 
+struct llama_model * llama_load_model_from_file(const char * path_model, struct llama_model_params params) {
+    return llama_load_model_from_file_internal(path_model, &params);
+}
+
```

Comment:
Is it necessary to add `llama_load_model_from_file_internal`?

Reply:
What do you think is the best way for GPT4All to fall back to CPU when a model is an unsupported arch or quant type, and inform the user of why the GPU couldn't be used? Should we compute the LLAMA_FTYPE_* value ourselves in the application?

Reply:
In my opinion it would be better to return an error in `llama_model_load` than to change the settings of the user without any notice. The application can disable GPU offloading after observing the error. However `llama_load_model_from_file_internal` doesn't seem relevant to any of this, it looks like a leftover from a previous change, but currently it does nothing.

Reply:
GPT4All currently uses llama_load_model_from_file_internal so it can read the changed ngl parameter, but I was afraid to expose that API upstream without further discussion.

Reply:
Let's:

- remove `llama_load_model_from_file_internal`
- in `llama_model_load` throw an error if the model and the params are not supported

In GPT4All, you can first try `llama_load_model_from_file` with `ngl > 0`. If it fails (i.e. returns `nullptr`), you can then try with `ngl == 0` and if it fails again - you do the same error handling that you do now

In the future, we will think about improving the API for such cases

Reply:
The problem with just printing errors to console is that GPT4All is a graphical application, and we have many users on Windows with no way to allocate a console to the program. And in general, users probably start GPT4All from a shortcut on their desktop, not the command line.

I can make this change for the sake of merging this PR, but I'll have to revert it on GPT4All's fork to preserve functionality. It is very important for us that users are given a specific reason why the GPU wasn't used - this change will create ambiguity between "GPU out of VRAM" and "unsupported quantization type or model architecture".

Reply:
We definitely need to do better error reporting in llama.cpp and ggml. I think it would be ok to keep this as a workaround until a error reporting framework is implemented in llama.cpp.

Reply:
I cleaned this up for now (llama_load_model_from_file_internal is gone). GPT4All will maintain modifications to this function until we can find a better way to communicate the error status to the caller.